goth/runner/__init__.py

```diff
@@ -240,32 +249,78 @@ async def __call__(
                 await self._enter()
                 yield self
             except asyncio.CancelledError:
-                self._cancellation_callback()
+                if self._cancellation_callback:
+                    self._cancellation_callback()
+                else:
+                    raise
             finally:
                 await self._exit()
         except TestFailure as err:
-            self._test_failure_callback(err)
+            if self._test_failure_callback:
+                self._test_failure_callback(err)
+            else:
+                logger.info("Runner stopped due to test failure")
 
-    async def _enter(self) -> None:
-        logger.info("Running test: %s", self._get_current_test_name())
+    @contextmanager
+    def _setup_file_logging(self) -> None:
+        """Set up logging to files in `self.log_dir`."""
 
-        self.base_log_dir.mkdir()
+        import goth
 
-        await self._exit_stack.enter_async_context(
-            run_compose_network(self._compose_manager, self.base_log_dir)
-        )
+        goth_logger = logging.getLogger(goth.__name__)
 
-        self._create_probes(self.base_log_dir)
+        import goth.api_monitor
+
+        api_monitor_logger = logging.getLogger(goth.api_monitor.__name__)
+
+        runner_handler = None
+        proxy_handler = None
+
+        try:
+            formatter = logging.Formatter(
+                fmt=LOGGING_CONFIG["formatters"]["date"]["format"],
+                datefmt=LOGGING_CONFIG["formatters"]["date"]["datefmt"],
+            )
+
+            # TODO: ensure the new files created here do not conflict with probe logs
+            runner_handler = logging.FileHandler(str(self.log_dir / "test.log"))
+            runner_handler.setLevel(logging.DEBUG)
+            runner_handler.setFormatter(formatter)
+            goth_logger.addHandler(runner_handler)
+
+            proxy_handler = logging.FileHandler(str(self.log_dir / "proxy.log"))
+            proxy_handler.setLevel(logging.DEBUG)
+            proxy_handler.setFormatter(formatter)
+            api_monitor_logger.addHandler(proxy_handler)
```

Comment:
I think this part can be moved to utility functions (e.g. `def add_proxy_handler(log_path: Path) -> Handler`) in `log.py` module. That would make them (slightly) more reusable and clean up this method.